### DISCUSSION TITLE
详解 Kubernetes Pod 的实现原理 · /kubernetes-pod

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#536] [Announcements]
https://draveness.me/kubernetes-pod

**Comment by ykfq** [comment]
非常深入。
问个题外话，请问作者的 ghost code injection 里面代码高亮部分是怎么配置的，看起来非常舒服。

**Comment by draveness** [comment]
> @ykfq
> 非常深入。
> 问个题外话，请问作者的 ghost code injection 里面代码高亮部分是怎么配置的，看起来非常舒服。

你可以看一下 prism.js

**Comment by ykfq** [comment]
好的，谢谢。

**Comment by goversion** [comment]
关于Pod 的创建，提到了沙盒这个关键字，我原来的理解是，沙盒即是容器，容器即是沙盒。请你讲一下，沙盒倒底是什么 ？

**Comment by draveness** [comment]
> @goversion
> 关于Pod 的创建，提到了沙盒这个关键字，我原来的理解是，沙盒即是容器，容器即是沙盒。请你讲一下，沙盒倒底是什么 ？

沙盒是 Pod 中一个特殊的容器 `k8s.gcr.io/pause:3.1`，仔细再看看吧

**Comment by Devying** [comment]
作者你好，容器是利用了Linux的 Namespace 和 Cgroup等实现了 隔离机制，那么Pod是怎么做到在Pod内部多个容器共享IPC，进程空间等资源的呢?

**Comment by draveness** [comment]
> 作者你好，容器是利用了Linux的 Namespace 和 Cgroup等实现了 隔离机制，那么Pod是怎么做到在Pod内部多个容器共享IPC，进程空间等资源的呢?

使用 Namespace 不是一定会隔离 IPC 和进程空间，它是一个选项

FYI：https://draveness.me/docker

**Comment by Devying** [comment]
刚我查到了资料，docker里面可以设置共享其他容器的 IPC，UTS，PID 等，猜测Pod是利用了这个特性。

**Comment by Formergg** [comment]
### 生命周期

在 Pod 被删除之前都会触发一个 PreStop 的钩子，其中的方法~之前~完成之后 Pod 才会被删除

---
其中的方法~之前~完成之后Pod才会被删除 -> 其中的方法完成之后Pod才会被删除



`m.executePreStopHook(pod, containerID, containerSpec, gracePeriod` 右边少了括号

---

*2020-07-10 UPDATES: 已修复*

**Comment by alpha-baby** [comment]
```golang
func (m *kubeGenericRuntimeManager) killContainersWithSyncResult(pod *v1.Pod, runningPod kubecontainer.Pod, gracePeriodOverride *int64) (syncResults []*kubecontainer.SyncResult) {
	containerResults := make(chan *kubecontainer.SyncResult, len(runningPod.Containers))

	for _, container := range runningPod.Containers {
		go func(container *kubecontainer.Container) {
			killContainerResult := kubecontainer.NewSyncResult(kubecontainer.KillContainer, container.Name)
			m.killContainer(pod, container.ID, container.Name, "Need to kill Pod", gracePeriodOverride)
			containerResults <- killContainerResult
		}(container)
	}
	close(containerResults)

	for containerResult := range containerResults {
		syncResults = append(syncResults, containerResult)
	}
	return
}
```
文中的这个代码是有问题的：
问题在：`close(containerResults)`运行后可能前面的for循环中的gotoutine还没结束发生：panic: send on closed channel

新代码中已经修复这个了：https://github.com/kubernetes/kubernetes/blob/8adcd7978edddaf64665504c75016ed227acf249/pkg/kubelet/kuberuntime/kuberuntime_container.go#L648

**Comment by koktlzz** [comment]
您好，读了您的文章真的获益匪浅，内容和配图都做得很棒。我有一个问题想问一下，请问源码中的InitContainer是否就是Pause容器呢？还望解答，谢谢！

**Comment by draveness** [comment]
> 您好，读了您的文章真的获益匪浅，内容和配图都做得很棒。我有一个问题想问一下，请问源码中的InitContainer是否就是Pause容器呢？还望解答，谢谢！

不是
